### PR TITLE
fix bug 1221110 - private, no-cache ban responses

### PR DESCRIPTION
**Pull request by groovecoder**
This adds `private, no-cache` to the response when a banned user's request is caught by the `BanMiddleware`.

Assuming our load-balancer follows standard `Cache-Control` implementation, this should stop it from caching these responses.

In theory, `no-cache` alone can do this; `private` is an extra indicator to the cache that the response is meant for a single user.

We can deploy and test this ourselves to see if it fixes the problem before getting WebOps involved debugging the load balancer.

### CODE REVIEW

**Comment by groovecoder** [comment]
Changing assignee to @robhudson

File: kuma/users/middleware.py [hunk 1]
```diff
@@ -17,8 +17,10 @@ def process_request(self, request):
...
-            return render(request,
-                          'users/user_banned.html',
-                          {'bans': bans,
-                           'path': request.path})
+            banned_response = render(request, 'users/user_banned.html', {
+                'bans': bans,
+                'path': request.path
+            })
+            banned_response['Cache-Control'] = 'private, no-cache'
+            return banned_response
```
Comment (robhudson) [inline kuma/users/middleware.py]: I'm not sure I like `private`. It suggests it's ok to cache this response in some situations (if the cache is private to this user).

Instead I would follow Django's `@never_cache` decorator and do what it does:
https://github.com/django/django/blob/master/django/views/decorators/cache.py#L57-L59

**Comment by robhudson** [comment]
r+ with the addition of a test for the cache header.

**Comment by jwhitlock** [comment]
I disagree with caching for the user either. They should be able to refresh the page and read the content, rather than the "you are a spammer" message.

**Comment by groovecoder** [comment]
Added test line for the response header, which I changed to just `no-cache`.

**Comment by groovecoder** [comment]
`never_cache` sets `Expires: <now>` ... I don't care which method we use. :neutral_face:

**Comment by robhudson** [comment]
I imagine Django has discovered some edge cases with caching headers over their years, which is why I like the idea of using `add_never_cache_headers` if the intention is to never cache this response.

**Comment by groovecoder** [comment]
Agree; updated to use `add_never_cache_headers` and to assert it was called in the test.

**Comment by groovecoder** [comment]
I don't understand why Travis is failing on this. The test passes on my local dev vm. @robhudson ?

**Comment by robhudson** [comment]
Tests pass for me locally also. I don't understand why either.

**Comment by groovecoder** [comment]
Updated test to assert `Expires` time is less than or equal to current time. (Also, timezones can go DIA:fire:)

File: kuma/users/tests/test_middleware.py [hunk 2]
```diff
@@ -24,3 +24,6 @@ def test_ban_middleware(self):
 
         resp = self.client.get('/')
         self.assertTemplateUsed(resp, 'users/user_banned.html')
+        expires_datetime = datetime.strptime(resp['Expires'],
+                                             '%a, %d %b %Y %H:%M:%S %Z')
+        assert expires_datetime.timetuple() <= gmtime()
```
Comment (robhudson) [inline kuma/users/tests/test_middleware.py]: `email.utils.parsedate` can parse these RFC 2822 dates giving something you can pass to `time.mktime` and compare it from there? Might save this ugly strptime format string?

Comment (robhudson) [inline kuma/users/tests/test_middleware.py]: I'm fine with this too if you'd prefer to run with it.

**Comment by robhudson** [comment]
r+wc

File: kuma/users/tests/test_middleware.py [hunk 3]
```diff
@@ -24,3 +24,4 @@ def test_ban_middleware(self):
 
         resp = self.client.get('/')
         self.assertTemplateUsed(resp, 'users/user_banned.html')
+        assert parsedate(resp['Expires']) <= gmtime()
```
Comment (robhudson) [inline kuma/users/tests/test_middleware.py]: We can also add `self.assertEqual(resp['cache-control'], 'max-age=0')`

Comment (groovecoder) [inline kuma/users/tests/test_middleware.py]: I thought about it, but it seems like django is changing `Cache-Control` behavior more readily than `Expires`, so I feel like this makes the test less susceptible to upstream logic changes. And the only reason I'm testing this is to make Travis pass ...

**Pull request merged**